micro-rdk-ffi/src/ffi/runtime.rs

```diff
@@ -0,0 +1,251 @@
+///
+/// The function will returns once the viam server is shutdown
+///
+/// # Safety
+/// `ctx` must be a valid pointer
+#[no_mangle]
+pub unsafe extern "C" fn viam_server_start(ctx: *mut viam_server_context) -> viam_code {
+    if ctx.is_null() {
+        return viam_code::VIAM_INVALID_ARG;
+    }
+
+    let ctx = unsafe { Box::from_raw(ctx) };
+
+    #[cfg(not(target_os = "espidf"))]
+    {
+        use log::LevelFilter;
+        env_logger::builder()
+            .format_timestamp(Some(env_logger::TimestampPrecision::Millis))
+            .filter_level(LevelFilter::Info)
+            .init()
+    }
+    #[cfg(target_os = "espidf")]
+    {
+        micro_rdk::esp32::esp_idf_svc::sys::link_patches();
+        micro_rdk::esp32::esp_idf_svc::log::EspLogger::initialize_default();
+        micro_rdk::esp32::esp_idf_svc::sys::esp!(unsafe {
+            micro_rdk::esp32::esp_idf_svc::sys::esp_vfs_eventfd_register(
+                &micro_rdk::esp32::esp_idf_svc::sys::esp_vfs_eventfd_config_t { max_fds: 5 },
+            )
+        })
+        .unwrap();
+    }
+
+    let max_connection = {
+        #[cfg(not(target_os = "espidf"))]
+        {
+            10
+        }
+        #[cfg(target_os = "espidf")]
+        {
+            use micro_rdk::esp32::esp_idf_svc::hal::sys::g_wifi_feature_caps;
+            use micro_rdk::esp32::esp_idf_svc::hal::sys::CONFIG_FEATURE_CACHE_TX_BUF_BIT;
+            if !g_spiram_ok {
+                log::info!("spiram not initialized disabling cache feature of the wifi driver");
+                g_wifi_feature_caps &= !(CONFIG_FEATURE_CACHE_TX_BUF_BIT as u64);
+                1
+            } else {
+                3
+            }
+        }
+    };
+
+    let repr = RobotRepresentation::WithRegistry(ctx.registry);
+
+    let network = {
+        #[cfg(not(target_os = "espidf"))]
+        {
+            use micro_rdk::common::conn::network::ExternallyManagedNetwork;
+            match local_ip_address::local_ip().expect("error parsing local IP") {
+                std::net::IpAddr::V4(ip) => ExternallyManagedNetwork::new(ip),
+                _ => panic!("oops expected ipv4"),
+            }
+        }
+        #[cfg(target_os = "espidf")]
+        {
+            use micro_rdk::esp32::conn::network::Esp32ExternallyManagedNetwork;
+            Esp32ExternallyManagedNetwork::default()
+        }
+    };
+
+    #[cfg(has_robot_config)]
+    {
+        use micro_rdk::common::provisioning::storage::RAMStorage;
+        let ram_storage = RAMStorage::new(
+            "",
+            "",
+            ROBOT_ID.expect("Provided build-time configuration failed to set `ROBOT_ID`"),
+            ROBOT_SECRET.expect("Provided build-time configuration failed to set `ROBOT_SECRET`"),
+        );
+        log::info!("Robot configuration information was provided at build time - bypassing Viam provisioning flow");
+        #[cfg(target_os = "espidf")]
+        use micro_rdk::esp32::entry::serve_web_with_external_network;
```

Comment:
The hardcoded choice of using an external network isn't generally correct for micro-rdk but that can be addressed later.

Reply:
I assume people using the ffi will be managing the network themselves. If they want to let micro rdk manage the network then they probably better off flushing the prebuilt binary